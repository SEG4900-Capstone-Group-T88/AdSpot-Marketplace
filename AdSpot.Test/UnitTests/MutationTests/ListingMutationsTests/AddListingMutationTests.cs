using AdSpot.Models;

namespace AdSpot.Test.UnitTests.ListingMutationsTests;

[Collection("adspot-inmemory-db")]
public class AddListingMutationTests
{
    private const string AddListingMutation = """
        mutation AddListing($input: AddListingInput!) {
            addListing(input: $input) {
                listing {
                    listingId
                }
                errors {
                    ... on Error {
                        __typename
                        message
                    }
                }
            }
        }
        """;

    [Fact]
    [Trait("Category", "Unit")]
    public async Task AddListingSuccessful()
    {
        var result = await TestServices.ExecuteRequestAsync(
            scope =>
            {
                var context = scope.ServiceProvider.GetRequiredService<AdSpotDbContext>();
                context.Connections.Add(
                    new Connection
                    {
                        UserId = TestDatabase.TestUser.UserId,
                        PlatformId = TestDatabase.Platforms.First().PlatformId,
                        Handle = "TestAccountHandle",
                        Token = "TestApiToken"
                    }
                );
                context.SaveChanges();
            },
            b =>
                b.SetDocument(AddListingMutation)
                    .SetVariableValues(
                        new Dictionary<string, object?>
                        {
                            {
                                "input",
                                new Dictionary<string, object?>
                                {
                                    { "listingTypeId", 1 },
                                    { "userId", TestDatabase.TestUser.UserId },
                                    { "price", (decimal)9.99 }
                                }
                            }
                        }.AsReadOnly()
                    )
        );

        result.MatchSnapshot();
    }

    [Fact]
    [Trait("Category", "Unit")]
    public async Task AddListingInvalidListingTypeId()
    {
        var result = await TestServices.ExecuteRequestAsync(b =>
            b.SetDocument(AddListingMutation)
                .SetVariableValues(
                    new Dictionary<string, object?>
                    {
                        {
                            "input",
                            new Dictionary<string, object?>
                            {
                                { "listingTypeId", -1 },
                                { "userId", TestDatabase.TestUser.UserId },
                                { "price", (decimal)9.99 }
                            }
                        }
                    }.AsReadOnly()
                )
        );

        result.MatchSnapshot();
    }

    [Fact]
    [Trait("Category", "Unit")]
    public async Task AddListingAccountHasNotBeenConnected()
    {
        var result = await TestServices.ExecuteRequestAsync(b =>
            b.SetDocument(AddListingMutation)
                .SetVariableValues(
                    new Dictionary<string, object?>
                    {
                        {
                            "input",
                            new Dictionary<string, object?>
                            {
                                { "listingTypeId", TestDatabase.ListingTypes.Last().ListingTypeId },
                                { "userId", TestDatabase.TestUser.UserId },
                                { "price", (decimal)9.99 }
                            }
                        }
                    }.AsReadOnly()
                )
        );

        result.MatchSnapshot();
    }
}
