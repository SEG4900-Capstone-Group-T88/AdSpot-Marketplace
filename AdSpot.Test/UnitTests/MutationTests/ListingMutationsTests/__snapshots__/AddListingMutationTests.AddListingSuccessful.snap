{
  "data": {
    "addListing": {
      "listing": {
        "listingId": 1
      },
      "errors": null
    }
  }
}